jans-cedarling/cedarling/src/jwt/test/with_validation/key_service.rs

```diff
@@ -0,0 +1,335 @@
+    });
+
+    // generate the signed token strings
+    let access_token = generate_token_using_claims(
+        &access_token_claims,
+        &EncodingKey {
+            key_id: "some_key_id_not_in_the_jwks".to_string(), // we set an non-existing `kid` for the access_token
+            key: encoding_keys[0].key.clone(),
+        },
+    );
+    let id_token = generate_token_using_claims(&id_token_claims, &encoding_keys[1]);
+    let userinfo_token = generate_token_using_claims(&userinfo_token_claims, &encoding_keys[0]);
+
+    // setup mock server responses for OpenID configuration and JWKS URIs
+    let openid_config_response = json!({
+        "issuer": server.url(),
+        "jwks_uri": &format!("{}/jwks", server.url()),
+    });
+    let openid_conf_mock = server
+        .mock("GET", "/.well-known/openid-configuration")
+        .with_status(200)
+        .with_header("content-type", "application/json")
+        .with_body(openid_config_response.to_string())
+        .create();
+    let jwks_uri_mock = server
+        .mock("GET", "/jwks")
+        .with_status(200)
+        .with_header("content-type", "application/json")
+        .with_body(jwks)
+        .create();
+
+    // initialize JwtService with validation enabled and ES256 as the supported algorithm
+    let jwt_service = JwtService::new_with_config(crate::jwt::JwtServiceConfig::WithValidation {
+        supported_algs: vec![Algorithm::ES256],
+        trusted_idps: vec![TrustedIssuer {
+            name: "some_idp".to_string(),
+            description: "some_desc".to_string(),
+            openid_configuration_endpoint: format!(
+                "{}/.well-known/openid-configuration",
+                server.url()
+            ),
+            token_metadata: None,
+        }],
+    });
+
+    // key service should fetch the jwks_uri on init
+    openid_conf_mock.assert();
+    // key service should fetch the jwks on init
+    jwks_uri_mock.assert();
+
+    // decode and validate the tokens
+    let decode_result = jwt_service
+        .decode_tokens::<serde_json::Value, serde_json::Value, serde_json::Value>(
+            &access_token,
+            &id_token,
+            &userinfo_token,
+        );
+
+    assert!(
+        matches!(
+            decode_result,
+            Err(jwt::Error::InvalidAccessToken(
+                jwt::decoding_strategy::Error::KeyService(jwt::decoding_strategy::key_service::Error::KeyNotFound(ref e))
+            )) if **e == *"some_key_id_not_in_the_jwks",
+        ),
+        "Expected decoding to fail due to not being able to find a key to validate `access_token`: {:?}",
+        decode_result
+    );
+    // key service should fetch the jwks again when it cant find the `kid` for the access_token
+    let jwks_uri_mock = jwks_uri_mock.expect(2);
+    jwks_uri_mock.assert();
+
+    // assert that there aren't any additional calls to the openid_config_uri
+    openid_conf_mock.assert();
+}
+
+#[test]
+#[should_panic]
```

Comment:
why we need use `#[should_panic]` ? why we can`t test in usual way?

Reply:
originally because of this
```rust
#[test]
#[should_panic]
/// Tests if [`JwtService::new_with_config`] panics if the JWKS cant be fetched
/// from the `jwks_uri`.
///
/// TODO: change this to check for an error instead of a panic once returning a `Result` for
/// the initialization for JwtService becomes supported
```
then i figured i just initilize the KeyService without the JwtService for the unit test here [7887aa5](https://github.com/JanssenProject/jans/pull/9999/commits/7887aa554baef4c42305efd8de520e8f66a22ea3)